### PR TITLE
store receipts and more rollup metadata to the db

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why is this change needed?

We need to store tx receipts and other metadata, so that we can expose them via RPC to be used by wallets (in a subsequent PR).
In a previous PR, the Merkle root of the receipts was added to the rollup header.

### What changes were made as part of this PR:

- functional PR 
- a bunch of accessor methods were copy-pasted from geth and slightly modified
- pass in the "receipts" to the storage function that saves the rollup, and store more metadata

### What are the key areas to look at
- storage - "SaveNewHead"

### CODE REVIEW

File: go/obscuronode/enclave/db/storage.go [hunk 1]
```diff
@@ -199,16 +199,25 @@ func (s *storageImpl) FetchBlockState(hash obscurocommon.L1RootHash) (*core.Bloc
...
-	}
+func (s *storageImpl) SaveNewHead(state *core.BlockState, rollup *core.Rollup, receipts []*types.Receipt) {
+	batch := s.db.NewBatch()
 
 	if state.FoundNewRollup {
-		s.StoreRollup(rollup)
+		obscurorawdb.WriteRollup(batch, rollup)
+		obscurorawdb.WriteCanonicalHash(batch, rollup.Hash(), rollup.NumberU64())
+		obscurorawdb.WriteHeadHeaderHash(batch, rollup.Hash())
+		obscurorawdb.WriteCanonicalHash(batch, rollup.Hash(), rollup.NumberU64())
```
Comment (joeldudley) [inline go/obscuronode/enclave/db/storage.go:209]: @tudor-malene Is this duplication of writing the canonical hash intentional?

Comment (tudor-malene) [reply go/obscuronode/enclave/db/storage.go:209]: no. good spot.

File: go/obscuronode/enclave/rawdb/accessors_indexes.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+// enabling hash based transaction and receipt lookups.
+func writeTxLookupEntry(db ethdb.KeyValueWriter, hash common.Hash, numberBytes []byte) {
+	if err := db.Put(txLookupKey(hash), numberBytes); err != nil {
+		log.Panic("Failed to store transaction lookup entry. Cause: %s", err)
+	}
+}
+
+// WriteTxLookupEntries is identical to WriteTxLookupEntry, but it works on
+// a list of hashes
+func WriteTxLookupEntries(db ethdb.KeyValueWriter, number uint64, hashes []common.Hash) {
```
Comment (joeldudley) [inline go/obscuronode/enclave/rawdb/accessors_indexes.go:40]: @tudor-malene A bunch of methods in this file are unused. Should we delete them?

File: go/obscuronode/enclave/rawdb/accessors_indexes.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/obscuronet/obscuro-playground/go/log"
+	"github.com/obscuronet/obscuro-playground/go/obscuronode/enclave/core"
+)
+
+// ReadTxLookupEntry retrieves the positional metadata associated with a transaction
```
Comment (joeldudley) [inline go/obscuronode/enclave/rawdb/accessors_indexes.go:15]: @tudor-malene I think we should make it clear using a comment in cases where we've copied functions directly from Geth.

File: go/obscuronode/enclave/rawdb/accessors_receipts.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/obscuronet/obscuro-playground/go/log"
+)
+
+// HasReceipts verifies the existence of all the transaction receipts belonging
```
Comment (joeldudley) [inline go/obscuronode/enclave/rawdb/accessors_receipts.go:14]: @tudor-malene As above, I think we should make it clear using a comment that this is code copied from Geth.

File: go/obscuronode/enclave/rawdb/accessors_indexes.go (hunk 3 above)
Comment (tudor-malene) [reply go/obscuronode/enclave/rawdb/accessors_indexes.go:15]: ok. I'll add a readme in the package. The entire thing is a slightly modified copy of the geth stuff

**Review by joeldudley — CHANGES_REQUESTED** [review]
Generally looks good. I think we should call out anywhere we're copy and pasting from Geth, to make it easier to update the code and understand where it came from later.

File: go/obscuronode/enclave/rawdb/accessors_indexes.go (hunk 2 above)
Comment (tudor-malene) [reply go/obscuronode/enclave/rawdb/accessors_indexes.go:40]: We need to use them eventually. Just need to figure out where to hook them in. I was planning to use this as part of this pr as well, but decided to push it early so you can use it.
I'll continue on it in a subsequent pr

**Comment by tudor-malene** [comment]
back to you @joeldudley .
Have a look at: https://github.com/obscuronet/go-obscuro/pull/281/files#diff-df16acc524215e3dba4ee7e622575de6f5382228a713e5034afb1a6b907c1f6b

**Review by joeldudley — APPROVED** [review]
LGTM

**Pull request merged**